File: C#/UNO-Game-API/lib/UnoGame.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;

namespace UnoGame
{
    public class UnoLib
    {
        public static void ShuffleDeck(List<Card> deckToShuffle)
        {
            deckToShuffle.Shuffle();
        }

        public static List<Card> GetHand(List<Card> deckToDealFrom, List<Card> currentGamePool)
        {
            if (deckToDealFrom.Count < 8)
            {
                MergePool(deckToDealFrom, currentGamePool);
                ShuffleDeck(deckToDealFrom);
            }

            List<Card> newHand = new List<Card>();
            for (int i = 0; i < 7 && deckToDealFrom.Count > 0; i++)
            {
                newHand.Add(deckToDealFrom[0]);
                deckToDealFrom.RemoveAt(0);
            }
            return newHand;
        }

        public static Card DrawCard(List<Card> deckToDrawFrom, List<Card> handToDrawTo, List<Card> currentGamePool)
        {
            if (deckToDrawFrom.Count < 8)
            {
                MergePool(deckToDrawFrom, currentGamePool);
                ShuffleDeck(deckToDrawFrom);
            }
            handToDrawTo.Add(deckToDrawFrom[0]);
            Card returnCard = deckToDrawFrom[0];
            deckToDrawFrom.RemoveAt(0);

            return returnCard;
        }

        public static void MergePool(List<Card> deckToMerge, List<Card> poolToMerge)
        {
            if (poolToMerge.Count == 0) return;

            Card topCard = poolToMerge[poolToMerge.Count - 1];
            poolToMerge.RemoveAt(poolToMerge.Count - 1);
            deckToMerge.AddRange(poolToMerge);
            poolToMerge.Clear();
            poolToMerge.Add(topCard);
        }

        public static void PlayCard(int cardIndex, List<Card> poolToPlay, List<Card> handToPlayFrom)
        {
            if (cardIndex < handToPlayFrom.Count)
            {
                poolToPlay.Add(handToPlayFrom[cardIndex]);
                handToPlayFrom.RemoveAt(cardIndex);
            }
        }

        public static List<Card> GetLegalPlays(List<Card> hand, Card topCard, int addTotal = 0)
        {
            List<Card> legalCards = new List<Card>();

            if (topCard == null)
                return hand;

            // Handle stacking logic if in a penalty state
            if (addTotal > 0)
            {
                foreach (var card in hand)
                {
                    if (topCard.value == 'T' && (card.value == 'T' || card.value == 'F')) // stack +2 or +4
                    {
                        legalCards.Add(card);
                    }
                    else if (topCard.value == 'F' && card.value == 'F') // only +4 on +4
                    {
                        legalCards.Add(card);
                    }
                }
                return legalCards;
            }

            // Normal play logic

            if (topCard.color == 'N')
            {
                return hand;
            }
            foreach (var card in hand)
            {
                if (card.value == topCard.value || card.color == topCard.color || card.color == 'N')
                {
                    legalCards.Add(card);
                }
            }

            return legalCards;
        }


        public static string GetTopCard(List<Card> poolInPlay)
        {
            return poolInPlay.Count > 0 ? poolInPlay[poolInPlay.Count - 1].GetCard() : " ";
        }

        public static Card GetTopCardC(List<Card> poolInPlay)
        {
            return poolInPlay.Count > 0 ? poolInPlay[poolInPlay.Count - 1] : null;
        }
        public static List<Card> InitBoard()
        {
            List<Card> deck = new List<Card>()
            {
                // Each color of card has 2 of type, but only ONE '0' card
                // Red cards
                new Card { value = '0', color = 'R'},
                new Card { value = '1', color = 'R'},
                new Card { value = '2', color = 'R'},
                new Card { value = '3', color = 'R'},
                new Card { value = '4', color = 'R'},
                new Card { value = '5', color = 'R'},
                new Card { value = '6', color = 'R'},
                new Card { value = '7', color = 'R'},
                new Card { value = '8', color = 'R'},
                new Card { value = '9', color = 'R'},
                new Card { value = 'T', color = 'R'}, // +2 card
                new Card { value = 'S', color = 'R'}, // Skip card
                new Card { value = 'R', color = 'R'}, // Reverse card

                new Card { value = '1', color = 'R'},
                new Card { value = '2', color = 'R'},
                new Card { value = '3', color = 'R'},
                new Card { value = '4', color = 'R'},
                new Card { value = '5', color = 'R'},
                new Card { value = '6', color = 'R'},
                new Card { value = '7', color = 'R'},
                new Card { value = '8', color = 'R'},
                new Card { value = '9', color = 'R'},
                new Card { value = 'T', color = 'R'}, // +2 card
                new Card { value = 'S', color = 'R'}, // Skip card
                new Card { value = 'R', color = 'R'}, // Reverse card

                // Blue cards
                new Card { value = '0', color = 'B'},
                new Card { value = '1', color = 'B'},
                new Card { value = '2', color = 'B'},
                new Card { value = '3', color = 'B'},
                new Card { value = '4', color = 'B'},
                new Card { value = '5', color = 'B'},
                new Card { value = '6', color = 'B'},
                new Card { value = '7', color = 'B'},
                new Card { value = '8', color = 'B'},
                new Card { value = '9', color = 'B'},
                new Card { value = 'T', color = 'B'}, // +2 card
                new Card { value = 'S', color = 'B'}, // Skip card
                new Card { value = 'R', color = 'B'}, // Reverse card

                new Card { value = '1', color = 'B'},
                new Card { value = '2', color = 'B'},
                new Card { value = '3', color = 'B'},
                new Card { value = '4', color = 'B'},
                new Card { value = '5', color = 'B'},
                new Card { value = '6', color = 'B'},
                new Card { value = '7', color = 'B'},
                new Card { value = '8', color = 'B'},
                new Card { value = '9', color = 'B'},
                new Card { value = 'T', color = 'B'}, // +2 card
                new Card { value = 'S', color = 'B'}, // Skip card
                new Card { value = 'R', color = 'B'}, // Reverse card

                // Yellow cards
                new Card { value = '0', color = 'Y'},
                new Card { value = '1', color = 'Y'},
                new Card { value = '2', color = 'Y'},
                new Card { value = '3', color = 'Y'},
                new Card { value = '4', color = 'Y'},
                new Card { value = '5', color = 'Y'},
                new Card { value = '6', color = 'Y'},
                new Card { value = '7', color = 'Y'},
                new Card { value = '8', color = 'Y'},
                new Card { value = '9', color = 'Y'},
                new Card { value = 'T', color = 'Y'}, // +2 card
                new Card { value = 'S', color = 'Y'}, // Skip card
                new Card { value = 'R', color = 'Y'}, // Reverse card

                new Card { value = '1', color = 'Y'},
                new Card { value = '2', color = 'Y'},
                new Card { value = '3', color = 'Y'},
                new Card { value = '4', color = 'Y'},
                new Card { value = '5', color = 'Y'},
                new Card { value = '6', color = 'Y'},
                new Card { value = '7', color = 'Y'},
                new Card { value = '8', color = 'Y'},
                new Card { value = '9', color = 'Y'},
                new Card { value = 'T', color = 'Y'}, // +2 card
                new Card { value = 'S', color = 'Y'}, // Skip card
                new Card { value = 'R', color = 'Y'}, // Reverse card

                // Green cards
                new Card { value = '0', color = 'G'},
                new Card { value = '1', color = 'G'},
                new Card { value = '2', color = 'G'},
                new Card { value = '3', color = 'G'},
                new Card { value = '4', color = 'G'},
                new Card { value = '5', color = 'G'},
                new Card { value = '6', color = 'G'},
                new Card { value = '7', color = 'G'},
                new Card { value = '8', color = 'G'},
                new Card { value = '9', color = 'G'},
                new Card { value = 'T', color = 'G'}, // +2 card
                new Card { value = 'S', color = 'G'}, // Skip card
                new Card { value = 'R', color = 'G'}, // Reverse card

                new Card { value = '1', color = 'G'},
                new Card { value = '2', color = 'G'},
                new Card { value = '3', color = 'G'},
                new Card { value = '4', color = 'G'},
                new Card { value = '5', color = 'G'},
                new Card { value = '6', color = 'G'},
                new Card { value = '7', color = 'G'},
                new Card { value = '8', color = 'G'},
                new Card { value = '9', color = 'G'},
                new Card { value = 'T', color = 'G'}, // +2 card
                new Card { value = 'S', color = 'G'}, // Skip card
                new Card { value = 'R', color = 'G'}, // Reverse card

                // Wild cards
                new Card { value = 'F', color = 'N'}, // +4 card
                new Card { value = 'F', color = 'N'}, // +4 card
                new Card { value = 'F', color = 'N'}, // +4 card
                new Card { value = 'F', color = 'N'}, // +4 card
                new Card { value = 'C', color = 'N'}, // Color change
                new Card { value = 'C', color = 'N'}, // Color change
                new Card { value = 'C', color = 'N'}, // Color change
                new Card { value = 'C', color = 'N'}, // Color change
            };

            return deck;
        }
    }

    public class Card
    {
        public char value { get; set; }
        public char color { get; set; }

        public string GetCard()
        {
            return value.ToString() + color.ToString();
        }
    }

    public static class ListExtensions
    {
        private static readonly Random rng = new Random();

        public static void Shuffle<T>(this IList<T> list)
        {
            int n = list.Count;
            while (n > 1)
            {
                n--;
                int k = rng.Next(n + 1);
                (list[n], list[k]) = (list[k], list[n]);
            }
        }
    }
}

File: Python/UNO-Game-API/UnoGame.py
import random

class UnoLib:
    @staticmethod
    def shuffle_deck(deck_to_shuffle):
        random.shuffle(deck_to_shuffle)

    @staticmethod
    def get_hand(deck_to_deal_from, current_game_pool):
        if len(deck_to_deal_from) < 8:
            UnoLib.merge_pool(deck_to_deal_from, current_game_pool)
            UnoLib.shuffle_deck(deck_to_deal_from)

        new_hand = []
        for i in range(min(7, len(deck_to_deal_from))):
            new_hand.append(deck_to_deal_from[0])
            deck_to_deal_from.pop(0)
        return new_hand

    @staticmethod
    def draw_card(deck_to_draw_from, hand_to_draw_to, current_game_pool):
        if len(deck_to_draw_from) < 8:
            UnoLib.merge_pool(deck_to_draw_from, current_game_pool)
            UnoLib.shuffle_deck(deck_to_draw_from)
        hand_to_draw_to.append(deck_to_draw_from[0])
        return_card = deck_to_draw_from[0]
        deck_to_draw_from.pop(0)

        return return_card

    @staticmethod
    def merge_pool(deck_to_merge, pool_to_merge):
        if len(pool_to_merge) == 0:
            return

        top_card = pool_to_merge.pop()
        deck_to_merge.extend(pool_to_merge)
        pool_to_merge.clear()
        pool_to_merge.append(top_card)

    @staticmethod
    def play_card(card_index, pool_to_play, hand_to_play_from):
        if card_index < len(hand_to_play_from):
            pool_to_play.append(hand_to_play_from[card_index])
            hand_to_play_from.pop(card_index)

    @staticmethod
    def get_legal_plays(hand, top_card, add_total=0):
        legal_cards = []

        if top_card is None:
            return hand

        # Handle stacking logic if in a penalty state
        if add_total > 0:
            for card in hand:
                if top_card.value == 'T' and (card.value == 'T' or card.value == 'F'):  # stack +2 or +4
                    legal_cards.append(card)
                elif top_card.value == 'F' and card.value == 'F':  # only +4 on +4
                    legal_cards.append(card)
            return legal_cards

        # Normal play logic
        if top_card.color == 'N':
            return hand
        for card in hand:
            if card.value == top_card.value or card.color == top_card.color or card.color == 'N':
                legal_cards.append(card)

        return legal_cards

    @staticmethod
    def get_top_card(pool_in_play):
        return pool_in_play[-1].get_card() if len(pool_in_play) > 0 else " "

    @staticmethod
    def get_top_card_c(pool_in_play):
        return pool_in_play[-1] if len(pool_in_play) > 0 else None

    @staticmethod
    def init_board():
        deck = [
            # Each color of card has 2 of type, but only ONE '0' card
            # Red cards
            Card('0', 'R'),
            Card('1', 'R'),
            Card('2', 'R'),
            Card('3', 'R'),
            Card('4', 'R'),
            Card('5', 'R'),
            Card('6', 'R'),
            Card('7', 'R'),
            Card('8', 'R'),
            Card('9', 'R'),
            Card('T', 'R'),  # +2 card
            Card('S', 'R'),  # Skip card
            Card('R', 'R'),  # Reverse card

            Card('1', 'R'),
            Card('2', 'R'),
            Card('3', 'R'),
            Card('4', 'R'),
            Card('5', 'R'),
            Card('6', 'R'),
            Card('7', 'R'),
            Card('8', 'R'),
            Card('9', 'R'),
            Card('T', 'R'),  # +2 card
            Card('S', 'R'),  # Skip card
            Card('R', 'R'),  # Reverse card

            # Blue cards
            Card('0', 'B'),
            Card('1', 'B'),
            Card('2', 'B'),
            Card('3', 'B'),
            Card('4', 'B'),
            Card('5', 'B'),
            Card('6', 'B'),
            Card('7', 'B'),
            Card('8', 'B'),
            Card('9', 'B'),
            Card('T', 'B'),  # +2 card
            Card('S', 'B'),  # Skip card
            Card('R', 'B'),  # Reverse card

            Card('1', 'B'),
            Card('2', 'B'),
            Card('3', 'B'),
            Card('4', 'B'),
            Card('5', 'B'),
            Card('6', 'B'),
            Card('7', 'B'),
            Card('8', 'B'),
            Card('9', 'B'),
            Card('T', 'B'),  # +2 card
            Card('S', 'B'),  # Skip card
            Card('R', 'B'),  # Reverse card

            # Yellow cards
            Card('0', 'Y'),
            Card('1', 'Y'),
            Card('2', 'Y'),
            Card('3', 'Y'),
            Card('4', 'Y'),
            Card('5', 'Y'),
            Card('6', 'Y'),
            Card('7', 'Y'),
            Card('8', 'Y'),
            Card('9', 'Y'),
            Card('T', 'Y'),  # +2 card
            Card('S', 'Y'),  # Skip card
            Card('R', 'Y'),  # Reverse card

            Card('1', 'Y'),
            Card('2', 'Y'),
            Card('3', 'Y'),
            Card('4', 'Y'),
            Card('5', 'Y'),
            Card('6', 'Y'),
            Card('7', 'Y'),
            Card('8', 'Y'),
            Card('9', 'Y'),
            Card('T', 'Y'),  # +2 card
            Card('S', 'Y'),  # Skip card
            Card('R', 'Y'),  # Reverse card

            # Green cards
            Card('0', 'G'),
            Card('1', 'G'),
            Card('2', 'G'),
            Card('3', 'G'),
            Card('4', 'G'),
            Card('5', 'G'),
            Card('6', 'G'),
            Card('7', 'G'),
            Card('8', 'G'),
            Card('9', 'G'),
            Card('T', 'G'),  # +2 card
            Card('S', 'G'),  # Skip card
            Card('R', 'G'),  # Reverse card

            Card('1', 'G'),
            Card('2', 'G'),
            Card('3', 'G'),
            Card('4', 'G'),
            Card('5', 'G'),
            Card('6', 'G'),
            Card('7', 'G'),
            Card('8', 'G'),
            Card('9', 'G'),
            Card('T', 'G'),  # +2 card
            Card('S', 'G'),  # Skip card
            Card('R', 'G'),  # Reverse card
        ]
        return deck


class Card:
    def __init__(self, value, color):
        self.value = value
        self.color = color

    def get_card(self):
        return str(self.value) + str(self.color)
